onnx/common/path.h

```diff
@@ -8,17 +8,58 @@
 #pragma once
 
 #include <string>
+#ifdef _WIN32
+#include <windows.h>
+#endif
 
 namespace ONNX_NAMESPACE {
 
 #ifdef _WIN32
 const std::string k_preferred_path_separator = "\\";
+const std::wstring w_k_preferred_path_separator = L"\\";
 #else // POSIX
 const std::string k_preferred_path_separator = "/";
 #endif
 
-std::string path_join(const std::string& origin, const std::string& append);
-void normalize_separator(std::string& path);
-std::string clean_relative_path(const std::string& path);
+template <typename STRING>
+STRING path_join(const STRING& origin, const STRING& append);
+
+template <typename CHAR>
+bool is_path_separator(CHAR c) {
+  // Windows accept / as path separator.
+  if (k_preferred_path_separator == "\\") {
+    return c == '\\' || c == '/';
+  }
+
+  return c == k_preferred_path_separator[0];
+}
+
+template <typename STRING>
+void normalize_separator(STRING& path) {
+  char preferred_sep = k_preferred_path_separator[0];
+  if (preferred_sep == '/') {
+    // Do nothing on linux.
+    return;
+  }
+
+  for (size_t i = 0; i < path.size(); i++) {
+    if (is_path_separator(path[i]) && path[i] != preferred_sep) {
+      path[i] = preferred_sep;
+    }
+  }
+}
+template <typename STRING>
+STRING clean_relative_path(const STRING& path);
+std::string clean_relative_path(const char* path);
```

Comment:
This part looks very confusing. Initially I thought the second one was specialization of the first one, Then I realized it missed "template" keyword. 
https://docs.microsoft.com/en-us/cpp/cpp/explicit-specialization-of-function-templates?view=msvc-170 

So, why for std::string you use template but for C style strings you use overloading? Why not using the same approach for both?

Reply:
Thanks for the pointer. I added it because we need to explicitly define a function for char[] input (directly giving raw string in double quote)

To prevent confusion, I also templatized this function and added comments. Now I added it as:
```
// std::string in template cannot be recognized as char[]
// Therefore explicitly define char[] to handle char[] input
template <typename CHAR, typename std::size_t N>
std::basic_string<CHAR> clean_relative_path(const CHAR (&path)[N]) {
  return clean_relative_path(std::basic_string<CHAR>(path));
}
```